Add cgroup memory controller

lebauce opened the pull request:
### What does this PR do?

Proactively call GC when memory used is aboved a specified threshold or when memory pressure events are sent by the kernel

### Motivation

Avoid OOM kills

lebauce committed:
Add cgroup memory controller

djmitche reviewed (changes requested):
I believe this needs an update to [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml)

djmitche commented:
..and perhaps a release note, since this is introducing an enhancement?

..and a rebase, since the dogstatsd size has increased recently :)

brycekahle commented on cmd/system-probe/utils/memory_monitor_linux.go:
```diff
@@ -0,0 +1,130 @@
+			}
+			defer memProfile.Close()
+
+			if err := pprof.WriteHeapProfile(memProfile); err != nil {
```
Should we log out the location of the memory profile file?

lebauce commented:
@djmitche Thanks for the review and sorry for the LONG delay. I addressed your comments and updated the config template. I'm not sure about the release notes though as this is mostly for troubleshooting purposes.

alai97 commented on pkg/config/config_template.yaml:
```diff
@@ -1386,6 +1386,38 @@ api_key:
+    ## memory usage is above the specified threshold.
+    ## Threshold can be either an absolute value - such as 500MB or 2GB -
+    ## or a percentage of the cgroup allocated memory such as 50%.
+    ## The action can be either:
```
Either implies one or the other (not three options).

```suggestion
    ## The action can be:
```

alai97 commented on pkg/config/config_template.yaml:
```diff
@@ -1386,6 +1386,38 @@ api_key:
+    ## @param pressure_levels - map of strings - optional
+    ## Pressure levels and the respective active actions to trigger when
+    ## memory usage reaches the specified level.
+    ## The pressure level can be either 'low', 'medium' or 'critical'.
+    ## The actions are the same than for thresholds (see above).
```
```suggestion
    ## The pressure level is 'low', 'medium', or 'critical'.
    ## The actions are the same for thresholds (see above).
```

alai97 reviewed (approved):
Looks good, just two copy nits. Should we maybe update `// Copyright 2016-2020 Datadog, Inc.` to `2016-2022 Datadog, Inc.`?

djmitche reviewed (commented):
The config template change looks OK for agent-core.

But, I want to be sure: is this something we'd like customers to know about and tweak on their own?  Or is it for internal / support use?  If the latter, then including it in the template may not be a good idea.

vboulineau commented:
@lebauce Would it make sense to add this feature to our own `cgroups` package: https://github.com/DataDog/datadog-agent/pull/9121?

lebauce commented:
@vboulineau Yes, it totally makes sense.

@djmitche This is for internal / support but it's hidden behind a boolean `InternalProfiling` false by default

@alai97 Applying your suggestions

djmitche reviewed (approved):
config template changes look good
